h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java

```diff
@@ -329,15 +343,47 @@ public void map(Chunk[] chks, NewChunk[] cress) {
           if (clf.vec().isString())
             for (int i = 0; i < clf._len; i++)
               cres.addNum(str_op(clf.atStr(lfstr, i), crt.atStr(rtstr, i)));
-          else
+          else if (categoricalOK() && clf.vec().isCategorical() && crt.vec().isCategorical()) {
```

Comment:
This check doesn't have to be done for every row, I would just do it in the initialization. For readability purposes, I would also pass a boolean flag to the MRTask that will indicate whether to use left or right for the mapping. This way you can keep the decision login (`clf.vec().domain().length > crt.vec().domain().length`) at one place outside of this task.

Reply:
These are done once before the MRTask is created (for categorical columns only). Then the values are red during iteration.